src/box/memtx_tree.cc

```diff
@@ -965,23 +1042,60 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
 	}
 
 /********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND START*********/
+	if (it->skip != 0) {
```

Comment:
This is out of scope of this PR, but:

Doesn't this kill the optimization? I understand there is no another way with our current isolation level, but I would argue this at least one example where one could benefit from weaker isolation levels.

Reply:
It does in case of a big number of concurrent statements. We've discussed offline with @alyapunov and @drewdzzz about this issue and it's been desided to prefer consistency over performance with minimal amount of changes until we have real MVCC users reporting about performance problems.

But I aggree with your point in general, MVCC-oblivious count would hurt much less in some cases.

Reply:
> But I aggree with your point in general, MVCC-oblivious count would hurt much less in some cases.

It’s not just about count, it’s also about select with offset.

Could you please open a ticket about allowing to lower the isolation level?

Reply:
Not sure I understood the your main concern. There're two aspects of this:
1. The line you refer to creates a global `GAP_COUNT` item, which slows down any concurrent inserts and deletes until the counting transaction is finished. Do you mean to introduce an ability to prevent count gap tracking? This way a transaction that once counted items in an index may have an inconsistent value received next time.
2. It's also possible (in theory), count raw tuples without traversing existing stories (which may be slow if we have a big amount of ones) to substract invisible ones. This way the transaction will potentially see uncommitted concurrent writes and subjects for garbage collection.

Reply:
Not sure I understood the idea. There're two aspects of this:
1. The line you refer to creates a global `GAP_COUNT` item, which slows down any concurrent inserts and deletes until the counted (selected with offset) transaction is finished. Do you mean to introduce an ability to prevent count gap tracking? This way a transaction that once counted items in an index may have an inconsistent value received next time. It's worse for select, where it can return different sets of values on each indentical request, for example:

   ```lua
   -- Index: [1, 3, 5, 7, 9]
   --[[TX1]]: select(nil, {offset = 3}) -- [7, 9]
   --[[TX2]]: insert(2)                 -- Now index: [1, 2, 3, 5, 7, 9].
   -- If TX1 doesn't create count gap item, TX2 doesn't conflict with it, so no read view for TX1.
   --[[TX1]]: select(nil, {offset = 3}) -- [5, 7, 9]
   ```

2. It's also possible (in theory), count raw tuples without traversing existing stories to substract invisible ones. This way the transaction will potentially count uncommitted concurrent writes and subjects for garbage collection. It may be OK for `index:len()`, but not for select with offset where we must only count the offset for visible tuples. The result for select is the same as above in this case.

Reply:
I was rather referring to 2, because I don’t like the additional traversing. But giving it a thought, I guess it doesn’t worsen the situation drastically, compared to the existing MVCC overheads.